[evohome] [new binding] Initial version of evohome binding for openhab2

Nebula83 opened the pull request:
This Pull-Request contains an initial version of a binding for the Honeywell evohome system.

Current supported functionality:
* Connecting using username and password
* Viewing and setting the system mode via the EvoTouch Gateway Thing
* Viewing the current set point and actual values of temperatures of all registered zones via the Zone thing.
* Overriding and cancellling an override of the temperature set point of all registered zones via the Zone thing.

signed-off-by: Jasper van Zuijlen <jvanzuijlen@gmail.com>
also-by: Neil Renaud <neil.renaud@gmail.com>

martinvw commented:
There are some commits which are not yours, for example the first one.

Maybe it's already a good moment to work on squashing the commits in a single commit with the correct sign-off and maybe some also-by. Can you already look into that?

Nebula83 commented:
Hi @martinvw, I'd be more than happy to do that. I have to admit that I am a total github n00b though. Could you refer me to a document as to how I go about this? I don't know what you mean by sign-off and also-by for instance. I also wouldn't know how to remove commits that are not mine but still keep in sync with the main repo? Sorry for my lack of skills 😅

martinvw commented:
The `Signed-off-by` is to certify that you wrote it or otherwise have the right to pass it on as an open-source, see http://docs.openhab.org/developers/contributing/contributing.html#sign-your-work. The also-by can be used if multiple persons collaborated on the code.

For squashing / rebasing see:
http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

We can also delay it till it is completely reviewed etc.

Nebula83 commented:
Thanks for the additional info, I'll have a look. Maybe close this PR until I have my history in order? Then I'll try again =)
One piece of advice maybe: I like to make small commits of my work in order to be able to retrace my steps. Would it be possible to do that and then to make a OH PR branch (2268-evohome) that has that concise history which I keep in sync with the master branch?

martinvw commented:
> Thanks for the additional info, I'll have a look. Maybe close this PR until I have my history in order? 

No just keep it open and visible, you can force push it later.

> I like to make small commits of my work in order to be able to retrace my steps.
You can do that we can squash later, in the end it will be merged to the master as a single commit.

Nebula83 commented:
So I spent a couple of hours the last days trying to get the history to be one commit (as I think is the question) but to no avail. git keeps throwing errors on that it cannot squash due to reasons I don't understand, sorry. Is there someone willing to help me out with this?

martinvw commented:
We will do it after the review, but please note there is a queue.

Nebula83 commented:
@martinvw Ok, thanks. I will continue to try and do it myself in the meantime. Problem with the queue, I have patience =)

foxy82 commented:
signed-off-by: Neil Renaud (github: foxy82)

Nebula83 commented:
Hi @martinvw, not to be impatient, but I didn't expect this review to be stale for a month. I don't think this review is really sensible anymore as there has been done a fair amount of work in the meantime. 

What is a practical approach for these pull requests? Only apply when it is fully done and bug free? I'm currently holding off development to not let the gap be too big between the current status and the status of the review which isn't really what I want...

martinvw commented:
Just keep developing. Also note that you named it WIP which normally means it's not yet in a mergable state, if you think it is please rename it. 

Note that everyone or at least most of us do this in our spare time, which is always limited. 

To give you an indication for the new bindings I am at the third page and you are on the top of the second, so I m getting closer but I'm not yet there.

Nebula83 commented:
Ok, got it then! Still new to this whole contribution thing. I didn't want to generate more work for you than necessary, that's all :-)

I didn't fully understand the meaning of the WIP tag. I'll process your remarks and see if I can get it into a mergable state. I myself do this in my sparetime too, so I know all about the limitedness of time ;-)

Thanks for the preliminary review!

Nebula83 commented:
@martinvw I processed you review comments (I think). I don't think I made it better though, as my build now fails? I tried to merge it, but that failed miserably. I can seems to get git merge the master into my branch. I strongly feel like recloning the official repo and just make one big commit of the current status of the binding. Any tips on how to make that happen?

martinvw commented:
You can delete your local branch, recreate it based on the current master and then add your big commit.

If you force push after that, you should be able to continue this PR with a clean start.

Ps: You can create a backup by creating a new branch based on the current one and pushing that to github.

Nebula83 commented:
Thanks @martinvw, I followed your directions, and I think it looks ok now.

martinvw commented:
```
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.openhab.binding:org.openhab.binding.evohome:2.1.1-SNAPSHOT (/home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.evohome/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.openhab.binding:org.openhab.binding.evohome:2.1.1-SNAPSHOT: Could not find artifact org.openhab.binding:pom:pom:2.1.0-SNAPSHOT in sonatype-snapshots (https://oss.sonatype.org/content/repositories/snapshots/) and 'parent.relativePath' points at wrong local POM @ line 7, column 10 -> [Help 2]
```

You should refer to 2.2.0-SNAPSHOT

martinvw commented:
@Nebula83 Please note that, because it's marked [WIP], I do and did not review this PR completely yet.

If it's finished and ready to review please update the title and mention me so I can remove the "Work in progress" label.

Nebula83 commented:
@martinvw Thanks for the hint on the 2.2.0. I didn't get why the build was failing, now I know and I fixed it.

As far as I'm concerned the PR is merge-worthy, please note that I removed the [WIP] tag from the title. If there is anything more I need to please tell me and I'll do so. 

I am aware that I do not have any unit tests at this point. If this is a blocking issue then I will focus on adding them to the review. I'd like to have some tests at strategic points anyway, but this is something I plan to work on after the merge.

Again, thanks for your patience!

Nebula83 commented:
@martinvw it's all set!

Nebula83 commented:
@martinvw I tried repushing my commits a couple of times, but the `Signed-off-by` check seens to always fail. Is there something I'm doing wrong?

martinvw commented:
> @martinvw I tried repushing my commits a couple of times, but the Signed-off-by check seens to always fail. Is there something I'm doing wrong?

Most likely you have configured a different email address in your git client then the one you are using for the sign off

Nebula83 commented:
@martinvw is that (or anything else) blocking this review?

Nebula83 commented on addons/binding/org.openhab.binding.evohome/ESH-INF/i18n/evohome_nl_NL.properties:
```diff
@@ -0,0 +1,11 @@
+# binding
+binding.evohome.name = Evohome binding
+binding.evohome.description = De evohome binding maakt het mogelijk om het Honeywell evohome systeem te besturen.
```
Taken windcentrale and plugwise as an example (the only bindings with Dutch translations in the repo), they seem to use "binding" as well. I'm usually not too happy with dutch translations of domain terms, so I'd like to propose to keep it like this.

Nebula83 commented on addons/binding/org.openhab.binding.evohome/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,52 @@
+            
+            <parameter name="password" type="text" required="true"
+                groupName="auth">
+                <label>Password</label>
```
Thanks! Where can I find the available contexts?

martinvw replied:
https://www.eclipse.org/smarthome/documentation/development/bindings/xml-reference.html#xml-structure-for-configuration-descriptions

Nebula83 commented on addons/binding/org.openhab.binding.evohome/README.md:
```diff
@@ -0,0 +1,47 @@
+
+The binding supports the following things:
+
+* evohome Account
```
I actually quite deliberately spelled it as 'evohome' since the product name is stylized like that. It it customary within openhab to spell product names with a capital?

martinvw replied:
You are so right, if you google honeywell's site spells it with a capital but if you click you see the real thing, all lowercase :-)

Nebula83 commented on addons/binding/org.openhab.binding.evohome/README.md:
```diff
@@ -0,0 +1,47 @@
+
+## Binding Configuration
+
+The evohome account needs to be configured with your username and password. Currently, only configuration through PaperUI is tested.
```
Will do

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/discovery/EvohomeDiscoveryService.java:
```diff
@@ -0,0 +1,130 @@
+ */
+public class EvohomeDiscoveryService extends AbstractDiscoveryService {
+    private Logger logger = LoggerFactory.getLogger(EvohomeDiscoveryService.class);
+    private static final int SEARCH_TIME = 2;
```
would I define that as `SEARCH_TIME_<UNIT_NAME_HERE>`?

martinvw replied:
👍

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/discovery/EvohomeDiscoveryService.java:
```diff
@@ -0,0 +1,130 @@
+                        discoverHeatingZones(controlSystem.getId(), controlSystem.getHeatingZones());
+                    }
+                }
+            } catch (Exception e) {
```
oops, overlooked that one

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/discovery/EvohomeDiscoveryService.java:
```diff
@@ -0,0 +1,130 @@
+        ThingUID thingUID = findThingUID(EvohomeBindingConstants.THING_TYPE_EVOHOME_DISPLAY.getId(), name);
+        Map<String, Object> properties = new HashMap<>();
+
+        //TODO NR: Doesn't the gateway also have a location? I don't have a dual location system but from the EvoHome UI
```
The TODO basically dictates a test with multiple locations. I don't know anyone with such a system, so testing it would be rather difficult. Would it be ok to put this in the docs? Currently the evohome binding only supports one location?

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/discovery/EvohomeDiscoveryService.java:
```diff
@@ -0,0 +1,130 @@
+        thingDiscovered(discoveryResult);
+    }
+
+    private ThingUID findThingUID(String thingType, String thingId) throws IllegalArgumentException {
```
ok, I think I let eclipse decide that for me? I'll remove it if I can

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeGatewayHandler.java:
```diff
@@ -0,0 +1,185 @@
+public class EvohomeGatewayHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(EvohomeGatewayHandler.class);
+    private EvohomeGatewayConfiguration configuration = null;
```
I didn't know I could that in java :-)

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeGatewayHandler.java:
```diff
@@ -0,0 +1,185 @@
+            apiClient = new EvohomeApiClientV2(configuration);
+
+            // Initialization can take a while, so kick if off on a separate thread
+            scheduler.schedule(new Runnable() {
```
Could you give me an example or a link? I don't know what you mean with java 8 notation 😊

martinvw replied:
Searching for `scheduler.schedule(` gives you some examples (at least in the master) XiaomiBridgeHandler line 104 for example

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeGatewayHandler.java:
```diff
@@ -0,0 +1,185 @@
+
+            updateGatewayStatus();
+            updateThings();
+        } catch (Exception e) {
```
Yeah, I do that a lot I see... Bad practice I know... I'll double check the binding in full...

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeGatewayHandler.java:
```diff
@@ -0,0 +1,185 @@
+            ThingHandler thingHandler = handler.getHandler();
+            if (thingHandler instanceof BaseEvohomeHandler) {
+                BaseEvohomeHandler moduleHandler = (BaseEvohomeHandler) thingHandler;
+                moduleHandler.update(apiClient);
```
I think I read that when you use a bridge, that entity such manage all the info. I think I try to adhere to the list you give. Do you see any exception to those rules?

martinvw replied:
There are always exceptions ;-) , but if you bump into one  just mention me then I will try to respond  :-) Or feel free to email me or contact me through the community forum

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeHeatingZoneHandler.java:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+
+        double temperature = zoneStatus.temperature.temperature;
```
Mostly to reduce line length. If that is undesirable, I'll refactor

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java:
```diff
@@ -0,0 +1,173 @@
+                if (out != null) {
+                    out = (TOut) new Gson().fromJson(reply, out.getClass());
+                }
+            }
```
I try to figure out what that means :-)

martinvw replied:
If you use the standard eclipse environment than you should be able to reformat your complete java src folder and it will have the right layout.

Nebula83 commented:
@martinvw  I just updated the PR with my new changes. Could you please do the full sweep?

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,252 @@
+            try {
+                httpClient.stop();
+            } catch (Exception e) {
+                logger.error("Could not stop http client.", e);
```
Given we are closing its not that important, just null it out (please do that anyway) and create a new one next time.

Please do not log to error in this case, debug is enough.

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,252 @@
+        boolean success = authenticateWithUsername();
+
+        // If the authentication succeeded, gather the basic intel as well
+        if (success == true) {
```
what about:

```java
if (success) {
```

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,252 @@
+    @Override
+    public void setTcsMode(String tcsId, String mode) {
+        String url = EvohomeApiConstants.URL_V2_BASE + EvohomeApiConstants.URL_V2_MODE;
+        url = String.format(url, tcsId);
```
Can't you string format it in one go?

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,252 @@
+        String url = EvohomeApiConstants.URL_V2_BASE + EvohomeApiConstants.URL_V2_ACCOUNT;
+
+        UserAccount userAccount = new UserAccount();
+        userAccount = apiAccess.doAuthenticatedRequest(HttpMethod.GET, url, null, null, userAccount);
```
Please return immediately

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,252 @@
+                + "&" + "Content-Type=application%2Fx-www-form-urlencoded%3B+charset%3Dutf-8&"
+                + "Connection=Keep-Alive";
+
+        HashMap<String, String> headers = new HashMap<String, String>();
```
Use the interface instead of the concrete class and use the diamond operator, ie:

```java
Map<String, String> headers = new HashMap<>();
```

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/request/HeatSetPoint.java:
```diff
@@ -0,0 +1,51 @@
+     *
+     * @param setPoint The target temperature to set the set point to
+     */
+    public HeatSetPoint(double setPoint) {
```
I personally would prefer (named) factory methods over this constructor, because now its not completely clear what will happen when you call which constructor, when using the factory method you have the name to clear that up.

Nebula83 replied:
I don't think a factory makes the design better, I chose to create a builder instead. I'll update the PR tomorrow if I can.

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/request/HeatSetPoint.java:
```diff
@@ -0,0 +1,51 @@
+    public double heatSetpointValue;
+
+    @SerializedName("SetpointMode")
+    public String setpointMode;
```
I'm not such a fan public fields not even in json objects :-)

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/request/Mode.java:
```diff
@@ -0,0 +1,42 @@
+ */
+public class Mode {
+
+    public Mode(String mode) {
```
See above

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/response/ActiveFault.java:
```diff
@@ -0,0 +1,26 @@
+    @SerializedName("faultType")
+    public String faultType;
+
+    //TODO Should be of time time, format: 2017-08-04T20:47:59.2030817
```
There is a TODO, what should be done and can it be completed before merging the PR?

Maybe you can parse it as an Instant?

Nebula83 replied:
copy paste remnant

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/request/Mode.java:
```diff
@@ -0,0 +1,42 @@
+    @SerializedName("SystemMode")
+    public String systemMode;
+
+    @SerializedName("TimeUntil")
```
Is the API really that inconistent in starting with lower or uppercase or is GSON case insensitive?

Nebula83 replied:
IDK, never tested it lowercase. I heavily rely on the python implementation - which in turn is sniffed from the official app -  and they use a Capital there. I can test to see if it works with lower as well, but I think this mimics the behavior of the official app better.

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/discovery/EvohomeDiscoveryService.java:
```diff
@@ -0,0 +1,138 @@
+ *
+ */
+public class EvohomeDiscoveryService extends AbstractDiscoveryService implements AccountStatusListener {
+    private Logger logger = LoggerFactory.getLogger(EvohomeDiscoveryService.class);
```
I prefer the loggers to be final

martinvw commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/discovery/EvohomeDiscoveryService.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    public void accountStatusChanged(ThingStatus status) {
+        if (status.equals(ThingStatus.ONLINE)) {
```
This is an enum AFAIK you can use ==

martinvw replied:
You also do that on line 89

Nebula83 replied:
Fixed and no? `if (bridge.getThing().getStatus() != ThingStatus.ONLINE)`

Nebula83 commented on addons/binding/org.openhab.binding.evohome/ESH-INF/thing/bridge.xml:
```diff
@@ -0,0 +1,37 @@
+			</parameter>
+			<parameter name="applicationId" type="text" required="true">
+				<label>Application Id</label>
+				<description>The application id to use (probably won't change).</description>
```
Well, this ID has been stable since I've started working on it. I'm doubting if it will ever change, but if it does, people could change it here. That way the binding wouldn't need an update. That said, I'm doubtful if it'll ever change. I can drop the statement in parenthesis maybe?

martinvw replied:
But is this a private ID only for you device or do you also share it with all other users?

Nebula83 replied:
It's the same for all users of this binding.

martinvw replied:
Then please remove it :-)

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java:
```diff
@@ -0,0 +1,176 @@
+     * @param out An instance of the return type, ignored when value is null
+     * @return The result of the request or null
+     */
+    @SuppressWarnings("unchecked")
```
See below

Nebula83 commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java:
```diff
@@ -0,0 +1,176 @@
+     */
+    @SuppressWarnings("unchecked")
+    public <TOut> TOut doRequest(HttpMethod method, String url, Map<String, String> headers, String requestData,
+            String contentType, TOut out) {
```
I changed the signature to `public <TOut> TOut doRequest(HttpMethod method, String url, Map<String, String> headers, String requestData, String contentType, Class<TOut> outClass, TOut out)`

This mitigates the need for unchecked since `retVal = new Gson().fromJson(reply, outClass);` now returns the actual TOut class and I don't have to hard-cast anymore.

Nebula83 commented:
@martinvw I updated the PR with the changes you requested. Could you check if it's suitable to merge now?

martinvw commented:
Please check the report of the static analysis, you can see it on travis or by just running maven clean install locally. After that please mention @kaikreuzer for a final review.

martinvw commented:
Please check the report of the static analysis, you can see it on travis or by just running maven clean install locally. After that please mention @kaikreuzer for a final review.

Nebula83 commented:
Yeah, I saw... I was just about to update my post :-)  I'll let you know when I fixed it!

martinvw commented:
At Kai thanks to some brilliant updates from GitHub I can no go longer assign issues / pr s from my tablet 😩

Nebula83 commented:
@martinvw @kaikreuzer I fixed the linting issue that where there, all should be ok now!

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+ org.openhab.binding.evohome,
+ org.openhab.binding.evohome.handler
+Import-Package: 
+ com.google.common.collect,
```
As you are using this only in a single line, could you completely remove this dependency? We are currently trying to get rid off the use of Guava.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/README.md:
```diff
@@ -0,0 +1,119 @@
+# evohome Binding
+
+This binding integrates the Honeywell evohome system. It uses your Honeywell Total Connect Comfort account to access your locations and heating zones.
```
Please add a line break after every sentence (it won't change the rendering, but will make diffing/reviewing easier).

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Binding Configuration
+
+The evohome account needs to be configured with your username and password. This can be done via PaperUI or things file.
```
The binding itself has no configuration parameters in your case (only the bridges/things), so you can remove the "Binding Configuration" section completely.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/README.md:
```diff
@@ -0,0 +1,119 @@
+
+## Thing Configuration
+
+Thing configuration is optional, it is easier to use discovery which will automatically add all your zones and displays to the inbox, once the account Thing is online.
```
In this section, you should simply describe the config parameters in a table (see e.g. [here](https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.gardena#account-configuration)) as they apply to Paper UI and textual configuration likewise - no need for examples, which should come below in the "full example" section.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/pom.xml:
```diff
@@ -0,0 +1,14 @@
+	<groupId>org.openhab.binding</groupId>
+	<artifactId>org.openhab.binding.evohome</artifactId>
+	<version>2.3.0-SNAPSHOT</version>
+	<name>Evohome Binding</name>
```
-> evohome

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,36 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Evohome Binding
```
-> evohome

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/AccountStatusListener.java:
```diff
@@ -0,0 +1,23 @@
+ */
+public interface AccountStatusListener {
+
+    public void accountStatusChanged(ThingStatus status);
```
please add a short JavaDoc

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeAccountBridgeHandler.java:
```diff
@@ -0,0 +1,263 @@
+                apiClient = new EvohomeApiClientV2(configuration);
+
+                // Initialization can take a while, so kick if off on a separate thread
+                scheduler.schedule(new Runnable() {
```
you could use lambda syntax: `scheduler.schedule(() -> {...}, 0, TimeUnit.SECONDS);`

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeAccountBridgeHandler.java:
```diff
@@ -0,0 +1,263 @@
+                        }
+                    }
+                }, 0, TimeUnit.SECONDS);
+            } catch (Exception e) {
```
This catch should be around the `apiClient` line only.

Nebula83 replied:
Not sure what you mean by that? I don't want to schedule the task when the EvohomeApiClientV2 throws and in `run` the apiClient is used on line 77 & 78?

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeAccountBridgeHandler.java:
```diff
@@ -0,0 +1,263 @@
+                    }
+                }, 0, TimeUnit.SECONDS);
+            } catch (Exception e) {
+                logger.error("Could not start API client", e);
```
Is it unexpected to ever get here, so that it is a bug that needs to be reported? Then logging ERROR is ok, but you might also directly want to throw an exception to signal that your handler initialisation failed.
No need to set the Thing status in that case.

Nebula83 replied:
Yes, that is true. Does that mean that I can remove the try...catch altogether?

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeAccountBridgeHandler.java:
```diff
@@ -0,0 +1,263 @@
+            updateThings();
+        } catch (Exception e) {
+            updateAccountStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            logger.warn("Failed to update installation status", e);
```
reduce to debug, you are already reporting the problem through the thing status.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/EvohomeHandlerFactory.java:
```diff
@@ -0,0 +1,88 @@
+
+    private void registerEvohomeDiscoveryService(EvohomeAccountBridgeHandler evohomeBridgeHandler) {
+        EvohomeDiscoveryService discoveryService = new EvohomeDiscoveryService(evohomeBridgeHandler);
+        discoveryServiceReg = bundleContext.registerService(DiscoveryService.class.getName(), discoveryService,
```
You are overwriting the previous value of discoveryServiceReg here. You must actually hold a Map of discoveryServiceRegs (see the [Hue binding as an example](https://github.com/eclipse/smarthome/blob/master/extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/HueThingHandlerFactory.java#L57)) as you have one discovery service per account.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java:
```diff
@@ -0,0 +1,198 @@
+                retVal = null;
+            }
+        } catch (TimeoutException | InterruptedException | ExecutionException e) {
+            logger.error("Error in handling request", e);
```
A timeout is clearly an expected path and it should merely set the Thing status to OFFLINE.
Logging errors must only be done if this is a bug situation that should be reported by users.

kaikreuzer replied:
You mean timeouts happen, even if everything works as expected? Sounds weird...
My point was that if you e.g. do not have a network connection, I would expect a TimeoutException here - and in that case, the Thing must be set to OFFLINE. Or is this already happening at some other place?
For the other two exceptions in that catch clause, you can choose error or warning as you like - assuming it is something that must normally never happen.

Nebula83 replied:
> You mean timeouts happen, even if everything works as expected?
Yeah, or my internet connection is unstable, but I see at least one timeout per day on requesting status. I could make it better by queueing up requests and implement a retry mechanism and throw a TimeoutException if happens `RetryCount` times, but that seems like something for a future version. For now, I don't really care if a status update fails. Then again, maybe I should solve it there then (only on update); let's go for that then.

kaikreuzer replied:
Setting the Thing to OFFLINE (and back to ONLINE after the next successful update) would at least give a hint to the user that something might not be fully stable&reliable.

Nebula83 replied:
Makes sense, I'll take that along

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java:
```diff
@@ -0,0 +1,198 @@
+
+        String json = null;
+        if (requestContainer != null) {
+            Gson gson = new GsonBuilder().create();
```
This is an expensive operation. You should keep the Gson instance as a field in the class and reuse it.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,258 @@
+    private final Logger logger = LoggerFactory.getLogger(EvohomeApiClientV2.class);
+
+    private final SslContextFactory sslContextFactory = new SslContextFactory();
+    private final HttpClient httpClient = new HttpClient(sslContextFactory);
```
You should not instantiate your own `HttpClient`, but rather use the shared client that can be obtained through the `HttpClientFactory` (see [here](https://www.eclipse.org/smarthome/documentation/development/bindings/dependencies.html#for-http-operations)). You can depend on that service in your handler factory and then reach the http client instance through the constructors, so that it ends up here.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClientV2.java:
```diff
@@ -0,0 +1,258 @@
+            locations = requestLocations();
+        } else {
+            apiAccess.setAuthentication(null);
+            logger.error("Authorization failed");
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/request/RequestBuilder.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public abstract class RequestBuilder<T> {
+
+    public abstract T Build();
```
change to `build()`.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/models/v2/request/RequestBuilder.java:
```diff
@@ -0,0 +1,21 @@
+ * @author Jasper van Zuijlen - Initial contribution
+ *
+ */
+public abstract class RequestBuilder<T> {
```
Couldn't that simply be an interface?

kaikreuzer commented:
Just as a heads-up: Tomorrow, 1pm would be deadline for getting this into the 2.3 release.

Nebula83 commented:
Tough luck, I just read this. 2.4 it is then.

Nebula83 commented:
Not yet, and I'm currently without a system and computer so it's going to have to wait...

Nebula83 commented:
@kaikreuzer @martinvw I updated the review with the changes requested. Please let me know if it meets the requirements now :-)

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/README.md:
```diff
@@ -0,0 +1,122 @@
+    sitemap evohome label="evohome Menu"
+    {
+        Frame label="evohome display" {
+            Selection label="[%s]" item=DemoMode mappings=[
```
The mappings should no longer be necessary, the UIs do that automatically.

Nebula83 replied:
Mine doesn't :-) Leave it there for completeness sake?

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/handler/EvohomeAccountBridgeHandler.java:
```diff
@@ -0,0 +1,275 @@
+            }
+
+            if (apiClient != null) {
+                // Initialization can take a while, so kick it off on a separate thread
```
... and as it can take a while, you should set the thing status to UNKNOWN here, before leaving the initialize method.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/EvohomeApiClient.java:
```diff
@@ -0,0 +1,256 @@
+ */
+public class EvohomeApiClient {
+
+    private static final String APPLICATION_ID = "b013aa26-9724-4dbd-8897-048b9aada249";
```
Where does this actually come from and is it ok to publish it publicly?

kaikreuzer commented on features/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -108,6 +108,11 @@
 
+    <feature name="openhab-binding-evohome" description="Evohome Binding" version="${project.version}">
+        <feature>openhab-runtime-base</feature>
+        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.evohome/${project.version}</bundle>
```
Please also add
```
         <feature>openhab-transport-http</feature>
```
here to make sure that the HTTP client is available.

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,60 @@
+	</channel-type>
+	<channel-type id="currentsetpoint">
+		<item-type>Number</item-type>
+		<label>Current set point</label>
```
Keep labels short and use capitalised strings. Would "Set Point" do here?

kaikreuzer replied:
I probably understand to little about how evohome works.
From openHAB perspective, any channel displays the "current" state by default, so this word should imho always be omitted. Furthermore, a "set point" - as the name suggest - is something to be "set"; what meaning does it have, when it is read-only? Shouldn't the set point and the override channel possibly be merged into one? I expect that if you override the set point, the "current set point" will show the value of your override, won't it?
Likewise you could get rid off the "cancel override" channel by e.g. defining that sending `NULL` to the setpoint cancels any override that might have been set.

Nebula83 replied:
Let me explain a bit more, since I'm not too happy with the current interface as well and you might have a more elegant solution to offer. Set point in this context refers to the desired temperature of a particular room. This can be set from the evohome in any number of ways and is is represented here as "current setpoint". Then there are two ways of overriding this setpoint : permanently (currently implemented) and temporary (not implemented yet). Finally, you want to cancel that override.

You example works very well for the current implementation, however with the future temporary override in mind; this is the best alternative I could come up with. Do you see a more openHABesque alternative?

kaikreuzer replied:
How temporary is temporary? Is it time based (i.e. for 1 hour) or can certain events cancel it (e.g. if temperature falls below 10°C)?
If it is purely time based, my take would be that the temporary override does not make much sense to be exposed to openHAB at all - users would rather use openHAB to decide when to cancel a permanent override again (e.g. sending a cancel after one hour in a rule).

Nebula83 replied:
It's time based. I can get on board with you reasoning. Let me see if I can set it up accordingly. Is sending `NULL` to a channel commonplace?

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,60 @@
+	</channel-type>
+	<channel-type id="permanentsetpoint">
+		<item-type>Number</item-type>
+		<label>Permanent set point override</label>
```
Maybe "Set Point Override" would do?

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,60 @@
+	</channel-type>
+	<channel-type id="cancelsetpoint">
+		<item-type>Switch</item-type>
+		<label>Cancel a set point override</label>
```
"Cancel Override"?

kaikreuzer commented on addons/binding/org.openhab.binding.evohome/ESH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,60 @@
+	</channel-type>
+	<channel-type id="setpointstatus">
+		<item-type>String</item-type>
+		<label>Set point status</label>
```
-> "Set Point Status"

Nebula83 commented:
@martinvw @kaikreuzer  I updated my PR again, I hope you can have look!

Nebula83 commented:
Hi @martinvw , @kaikreuzer. I fixed the merge conflicts, so it should be in sync again.

Nebula83 commented:
@kaikreuzer wow! Great news! I fixed the last bit, so all should be good to go.
